routing/triert/table.go

```diff
@@ -0,0 +1,227 @@
+		keyNodesNext[k] = v
+	}
+	keyNodesNext[string(kk)] = node
+
+	rt.keyNodes.Store(keyNodesNext)
+	rt.keys.Store(keysNext)
+	return true, nil
+}
+
+// RemoveKey tries to remove a peer identified by its Kademlia key from the
+// routing table
+func (rt *TrieRT) RemoveKey(ctx context.Context, kk key.KadKey) (bool, error) {
+	if kk.Size() != rt.self.Size() {
+		return false, ErrKeyWrongLength
+	}
+	rt.keymu.Lock()
+	defer rt.keymu.Unlock()
+	// load the old trie, derive a mutated variant and store it in place of the original
+	keys := rt.keys.Load().(*trie.Trie)
+	keysNext := trie.Remove(keys, tkey.Key(kk))
+
+	// if trie is unchanged then we didn't remove the key
+	if keysNext == keys {
+		return false, nil
+	}
+
+	// make a copy of keyNodes without removed key
+	// could avoid this if we held key/nodeid tuple in the trie
+	keyNodes := rt.keyNodes.Load().(map[string]address.NodeID)
+	if _, exists := keyNodes[string(kk)]; exists {
+		keyNodesNext := make(map[string]address.NodeID, len(keyNodes))
+		for k, v := range keyNodes {
+			if k == string(kk) {
+				continue
+			}
+			keyNodesNext[k] = v
+		}
+		rt.keyNodes.Store(keyNodesNext)
+	}
+
+	rt.keys.Store(keysNext)
+	return true, nil
+}
+
+// NearestPeers returns the n closest peers to a given key
+func (rt *TrieRT) NearestPeers(ctx context.Context, kk key.KadKey, n int) ([]address.NodeID, error) {
+	if kk.Size() != rt.self.Size() {
+		return nil, ErrKeyWrongLength
+	}
+
+	keys := rt.keys.Load().(*trie.Trie)
+	closestKeys, err := closestAtDepth(ctx, tkey.Key(kk), keys, 0, n, make([]tkey.Key, 0, n))
+	if err != nil {
+		return nil, err
+	}
+	keyNodes := rt.keyNodes.Load().(map[string]address.NodeID)
+
+	nodes := make([]address.NodeID, 0, len(closestKeys))
+	for _, c := range closestKeys {
+		if id, ok := keyNodes[string(key.KadKey(c))]; ok {
+			nodes = append(nodes, id)
+		}
+	}
+
+	return nodes, nil
+}
+
+func closestAtDepth(ctx context.Context, k tkey.Key, t *trie.Trie, depth int, n int, found []tkey.Key) ([]tkey.Key, error) {
+	if ctx.Err() != nil {
+		return found, ctx.Err()
+	}
+
+	// If we've already found enough peers, abort.
+	if n == len(found) {
+		return found, nil
+	}
+
+	// Find the closest direction.
+	dir := k.BitAt(depth)
+	var chosenDir byte
+	if t.Branch[dir] != nil {
+		// There are peers in the "closer" direction.
+		chosenDir = dir
+	} else if t.Branch[1-dir] != nil {
+		// There are peers in the "less closer" direction.
+		chosenDir = 1 - dir
+	} else if t.Key != nil {
+		// We've found a leaf
+		return append(found, t.Key), nil
+	} else {
+		// We've found an empty node?
+		return found, nil
+	}
+
+	var err error
+	// Add peers from the closest direction first, then from the other direction.
+	found, err = closestAtDepth(ctx, k, t.Branch[chosenDir], depth+1, n, found)
+	if err != nil {
+		return found, err
+	}
+
+	return closestAtDepth(ctx, k, t.Branch[1-chosenDir], depth+1, n, found)
+}
+
+func (rt *TrieRT) Find(ctx context.Context, kk key.KadKey) (address.NodeID, error) {
```

Comment:
It could be used to verify if a peer is already in the routing table or not, e.g in the context of https://github.com/libp2p/go-libp2p-kad-dht/issues/811 where we send a kademlia request to the remote peer before adding it to the routing table.

In go-libp2p-kbucket this logic is implemented in https://github.com/libp2p/go-libp2p-kbucket/blob/6d85d4e63aa3b8e04b9b55e1de64be8a52532f21/table.go#L119. So there is no use case now (except testing), but it may be useful in the future.

Reply:
It's used in tests and potentially in other diagnostics / consistency checking. It's not part of the Table interface.

Reply:
If it's possible we could un-export it.

Reply:
I don't see the need to unexport it. It's a useful function for code that want to directly query the table